synapse/storage/database.py

```diff
@@ -135,46 +132,61 @@ class LoggingTransaction(object):
     ]
 
     def __init__(
-        self, txn, name, database_engine, after_callbacks=None, exception_callbacks=None
+        self,
+        txn: Cursor,
+        name: str,
+        database_engine: BaseDatabaseEngine,
+        after_callbacks: Optional[List[_CallbackListEntry]] = None,
+        exception_callbacks: Optional[List[_CallbackListEntry]] = None,
     ):
-        object.__setattr__(self, "txn", txn)
-        object.__setattr__(self, "name", name)
-        object.__setattr__(self, "database_engine", database_engine)
-        object.__setattr__(self, "after_callbacks", after_callbacks)
-        object.__setattr__(self, "exception_callbacks", exception_callbacks)
+        self.txn = txn
+        self.name = name
+        self.database_engine = database_engine
+        self.after_callbacks = after_callbacks
+        self.exception_callbacks = exception_callbacks
 
-    def call_after(self, callback, *args, **kwargs):
+    def call_after(self, callback: "Callable[..., None]", *args, **kwargs):
         """Call the given callback on the main twisted thread after the
         transaction has finished. Used to invalidate the caches on the
         correct thread.
         """
+        assert self.after_callbacks is not None
```

Comment:
Not sure I understand why we're raising an exception here if `self.after_callbacks` is None and not assigning a new list in that case.

Reply:
well, it's up to the thing that constructs the `LoggingTransaction` to make sure that anything that gets added to `after_callbacks` actually gets called. So if we just create an `after_callbacks` ourselves, then the callback will just get silently dropped.

In short, if anything calls `call_after` when `after_callbacks is None`, that means there's a programming error.

No doubt we could factor out the code to run `after_callbacks` to avoid the problem, but this PR was about annotating the current impl rather than trying to change too much.

have added some comments which might help explain.

Reply:
Right, fair enough, fwiw I think the comment is enough.